### PR TITLE
Create database configuration and docker setup

### PR DESCRIPTION
**Pull request by Nipa-Code**
Commit includes database configuration with docker and optionally with your own host. Create `req.py` to make requests to endpoints that require database, makes testing the database easier.

### CODE REVIEW

File: shark/endpoints/testing.py [hunk 1]
```diff
@@ -1,12 +1,51 @@
...
+from shark.utils.database import engine
 
 router = APIRouter()
 
+models.Base.metadata.create_all(bind=engine)
+
 
 @router.get("/testing/")
-async def test(request: Request):
+async def test(request: Request, db):
```
Comment (CaedenPH) [inline shark/endpoints/testing.py:15]: What is `db`?

File: shark/endpoints/testing.py [hunk 2]
```diff
@@ -1,12 +1,51 @@
...
     return {"message": f"Wow, test succesful! {name}"}
+
+
+@router.post("/book/", response_model=schemas.Book)
+async def post_book(book: schemas.Book, db: Session = Depends(get_db)):
+    db_book = models.Book(
+        title=book.title, rating=book.rating, author_id=book.author_id
+    )
+    db.add(db_book)
+    db.commit()
```
Comment (CaedenPH) [inline shark/endpoints/testing.py:30]: Should the database not be asynchronous?

**Review by CaedenPH — CHANGES_REQUESTED** [review]
- [ ] Without an asynchronous database wrapper, the code will block with larger requests

File: shark/endpoints/testing.py (hunk 1 above)
Comment (Nipa-Code) [reply shark/endpoints/testing.py:15]: thing that should not be there, should of be get rid of

File: shark/endpoints/testing.py (hunk 2 above)
Comment (Nipa-Code) [reply shark/endpoints/testing.py:30]: It can be both, if I have understood the docs at all correctly it should be asynchronous atm.

**Comment by CaedenPH** [comment]
You also provide no data validation checks, ie is the author name longer than a certain length, etc.

For a robust route, you need:
  - range check
  - length check
  - type check 
  - presence check

File: req.py [hunk 3]
```diff
@@ -0,0 +1,23 @@
+# example file to make requests to the API endpoints that require DB
```
Comment (CaedenPH) [inline req.py:1]: Move this file to a `/tests/` directory

File: Dockerfile [hunk 4]
```diff
@@ -0,0 +1,7 @@
+FROM python:3.9
```
Comment (CaedenPH) [inline Dockerfile:1]: Why not latest?

File: Dockerfile [hunk 5]
```diff
@@ -0,0 +1,7 @@
...
+WORKDIR /shark
+COPY requirements.txt requirements.txt
+RUN /usr/local/bin/python -m pip install --upgrade pip && pip install --no-cache-dir --upgrade -r requirements.txt
```
Comment (CaedenPH) [inline Dockerfile:4]: Why are you running `python`/`pip` from 3 different places, `/usr/local/bin/python`, `pip` and `python3`

File: .env_template [hunk 6]
```diff
@@ -0,0 +1,7 @@
+DB_URL = ""
```
Comment (CaedenPH) [inline .env_template:1]: You could just name this file `.env.template`

File: requirements.txt [hunk 7]
```diff
@@ -2,4 +2,9 @@ fastapi == 0.85.0
...
 uvicorn == 0.18.3
 jinja2 == 3.1.2
-black == 22.8.0
\ No newline at end of file
+black == 22.8.0
+flake8 == 5.0.4
+SQLAlchemy == 1.4.41
+psycopg2 == 2.9.3
+python-decouple == 3.6
+requests == 2.28.1
```
Comment (CaedenPH) [inline requirements.txt:10]:
```suggestion
```
`requests` shouldn't be a dependency, it's a dev requirement

File: requirements.txt [hunk 8]
```diff
@@ -2,4 +2,9 @@ fastapi == 0.85.0
 loguru == 0.6.0
 uvicorn == 0.18.3
 jinja2 == 3.1.2
-black == 22.8.0
\ No newline at end of file
+black == 22.8.0
+flake8 == 5.0.4
```
Comment (CaedenPH) [inline requirements.txt:6]:
```suggestion
```

Dev requirement

File: requirements.txt [hunk 9]
```diff
@@ -2,4 +2,9 @@ fastapi == 0.85.0
 loguru == 0.6.0
 uvicorn == 0.18.3
 jinja2 == 3.1.2
-black == 22.8.0
\ No newline at end of file
+black == 22.8.0
```
Comment (CaedenPH) [inline requirements.txt:5]:
```suggestion
```
Dev requirement

File: Dockerfile [hunk 10]
```diff
@@ -0,0 +1,7 @@
...
+EXPOSE 5001
+COPY . .
```
Comment (CaedenPH) [inline Dockerfile:6]: I'm not a docker expert, what does this do?

File: shark/constants.py [hunk 11]
```diff
@@ -5,4 +5,4 @@ class Config:
     TEMPLATES = Jinja2Templates(directory="shark/templates")
     TITLE: str = "Admin Panel"
     DESCRIPTION: str = "Fancy way to communicate with your API and do testing"
-    VERSION: str = "0.1"
\ No newline at end of file
+    VERSION: str = "0.1"
```
Comment (CaedenPH) [inline shark/constants.py:8]:
```suggestion
```

All these 'constants' should go in the `__init__.py` as dunders; 
`__version__ = "0.2"` - you may also want to increment the version

File: shark/endpoints/testing.py [hunk 12]
```diff
@@ -1,12 +1,51 @@
-from fastapi import APIRouter, Request
+from fastapi import APIRouter, Request, Depends
+
+from sqlalchemy.orm import Session
+
+from shark.utils.crud import get_db
+from shark.utils import models, schemas
+from shark.utils.database import engine
```
Comment (CaedenPH) [inline shark/endpoints/testing.py:7]:
```suggestion
from .utils import models, schemas
from .utils.crud import get_db
from .utils.database import engine
```

These should be relative imports.

`utils` dif should also have an `__init__` importing relative functions/classes

File: Dockerfile (hunk 4 above)
Comment (Nipa-Code) [reply Dockerfile:1]: it's copied from my other API so I haven't cared yet to update all the values, eg. the DB is still using version (12?) aand the newest is 14 or 15. but it works

File: requirements.txt (hunk 7 above)
Comment (Nipa-Code) [reply requirements.txt:10]: it will have use, assuming we are not going to use standard libs in the future

File: requirements.txt (hunk 8 above)
Comment (Nipa-Code) [reply requirements.txt:6]: will be used for testing in the future (probably)

File: requirements.txt (hunk 9 above)
Comment (Nipa-Code) [reply requirements.txt:5]: will be used for testing in the future (probably)

File: shark/constants.py (hunk 11 above)
Comment (Nipa-Code) [reply shark/constants.py:8]: Every PR = new version? never really used `version` stuff before

File: Dockerfile (hunk 4 above)
Comment (CaedenPH) [reply Dockerfile:1]: Okay

File: Dockerfile (hunk 5 above)
Comment (Nipa-Code) [reply Dockerfile:4]: if it works, don't touch it

**Review by CaedenPH — APPROVED** [review]

**Pull request merged**